### ISSUE TITLE
Add support for Monix subscriptions

### ISSUE THREAD

**Issue by jvican** [ghostdogpr/caliban#212]
Hi! I'm considering using caliban to do GraphQL in a Monix-based application. I've seen you have support for cats via enrichments such as `executeAsync` but I haven't seen there's support to use `Observable/Observer` when implementing subscriptions.
  
I'd like to propose that caliban provides a new module that, in short, allows me to do something like this example:
  
```scala
case class Subscriptions(deletedCharacter: Observable[Character])
val subscriptions = Subscriptions(???)
```

**Comment by ghostdogpr** [comment]
@jvican Hi! (big fan of your work on bloop so happy to see you here 😄)

This sounds totally doable. We could have a module `caliban-monix`, similar to the `caliban-cats` that would provide a `Schema` for `Observable` (and any other needed type).
```scala
def observableSchema[R, A](implicit ev: Schema[R, A]): Schema[R, Observable[A]] =
    new Schema[R, Observable[A]] {
      override def optional: Boolean                        = ev.optional
      override def toType(isInput: Boolean = false): __Type = ev.toType(isInput)
      override def resolve(value: Observable[A]): Step[R]   = StreamStep(???)
    }
```
The only missing bit is how to convert an `Observable` to a `ZStream`. I have never used Monix before but from the docs it seems to be push-based, so we could push received items in a `ZQueue` and use `ZStream.fromQueue`.

We could also offer a native support for Monix `Task` in that module to avoid the cats-effect indirection.

Would you like to work on it? If not, I'm happy to have a look at it, I just need to learn some Monix basics but that shouldn't be too hard.

**Comment by ghostdogpr** [comment]
We could even use [zio-interop-reactive-streams](https://github.com/zio/interop-reactive-streams) since `Observable` is compatible with Reactive Streams.

**Comment by jvican** [comment]
I think a `caliban-monix` module would make a lot of sense :smile: Ideally, we'd have native support for Monix and avoid conversions between the pull-based model and the push-based model, which would hurt performance, especially sensitive when streaming.
  
I don't currently have the time to look at this but it isn't super time pressing. I imagine it would take me a bit to make this contribution and navigate the codebase.
  
If you can have a look at this, I'm happy to help out with Monix concepts/basics via Gitter or some async chat such as Signal or Telegram.
  
Also, a somewhat related question: how tied is Caliban to use ZIO under the hood? Would it be possible to change Caliban to make it work with Monix natively?
  
I'm currently fine using the interop modules provided for cats but I'm concerned about overall application size. I need to bundle the application in a big docker image where the total size must be minimized as much as possible.
  
P.S. Thanks for creating this library, it's great

**Comment by ghostdogpr** [comment]
OK, I will give it a try and let you know if I have any difficulties.

To answer your questions: yes, Caliban is pretty much tied to ZIO. That is because it takes full benefit of ZIO core features, namely the explicit error type and the embedded environment. That way, ZIO users can have the "full" experience while CE/Monix users only use a subset (no environment, E = Throwable). If the library used Cats Effect instead, the experience for ZIO users would be degraded. Another reason is that I wanted the library to be super accessible, usable without understanding type classes or importing implicits.

About dependencies: Caliban only depend on Magnolia, Fastparse (both are very small), ZIO and ZIO Streams. ZIO itself has zero dependency so you only get its own weight: 1.4MB for ZIO and 0.5MB for streams. It doesn't bring bigger libraries like Cats or Scalaz into the classpath (but I think Monix brings you Cats anyway).

**Comment by ghostdogpr** [comment]
Made PR #214 for this. It seems to work fine on a simple example with both `Task` and `Observable`. I'm not sure if the conversion ZIO <-> Monix is the most efficient possible so any feedback is welcome.

**Comment by ghostdogpr** [comment]
Released version 0.5.2 with this!
